src/sentry/auth/providers/saml2.py

```diff
@@ -1,65 +1,134 @@
 
 from onelogin.saml2.auth import OneLogin_Saml2_Auth, OneLogin_Saml2_Settings
 
 from sentry.auth import Provider, AuthView
-from sentry.models import AuthProvider, Organization, OrganizationStatus
+from sentry.auth.exceptions import IdentityNotValid
+from sentry.models import (AuthProvider, Organization, OrganizationStatus,
+                           User, UserEmail)
 from sentry.utils.http import absolute_uri
+from sentry.utils.auth import login, get_login_redirect
+
+
+def get_provider(organization_slug):
+    try:
+        organization = Organization.objects.get(
+            slug=organization_slug
+        )
+    except Organization.DoesNotExist:
+        return HttpResponseRedirect(reverse('sentry-login'))
+
+    if organization.status != OrganizationStatus.VISIBLE:
+        return HttpResponseRedirect(reverse('sentry-login'))
+
+    try:
+        auth_provider = AuthProvider.objects.get(
+            organization=organization
+        )
+        return auth_provider.get_provider()
+    except AuthProvider.DoesNotExist:
+        return HttpResponseRedirect(reverse('sentry-login'))
 
 
 class SAML2LoginView(AuthView):
     def dispatch(self, request, helper):
-        if 'SAMLResponse' in request.POST.keys():
-            return helper.next_step()
-        saml_config = helper.provider.build_saml_config(helper.organization.slug)
-        auth = helper.provider.build_auth(request, saml_config)
+        provider = helper.provider
+        saml_config = provider.build_saml_config(helper.organization.slug)
+        auth = provider.build_auth(request, saml_config)
         return self.redirect(auth.login())
 
 
 class SAML2ACSView(AuthView):
-    def dispatch(self, request, helper):
-        saml_config = helper.provider.build_saml_config(helper.organization.slug)
-        auth = helper.provider.build_auth(request, saml_config)
+    @method_decorator(csrf_exempt)
+    def dispatch(self, request, organization_slug):
+        provider = get_provider(organization_slug)
+        organization = Organization.objects.get(slug=organization_slug)
+        saml_config = provider.build_saml_config(organization_slug)
+
+        auth = provider.build_auth(request, saml_config)
         auth.process_response()
         errors = auth.get_errors()
         if errors:
             error_reason = auth.get_last_error_reason()
             raise IdentityNotValid(error_reason)
 
-        data = {
-            'attributes': auth.get_attributes(),
-            'nameid': auth.get_nameid(),
-            'session_index': auth.get_session_index()
-        }
+        attributes = auth.get_attributes()
+        nameid = auth.get_nameid()
 
-        helper.bind_state('data', data)
-        return helper.next_step()
-        #return HttpResponse("ACS endpoint")
+        email = self.retrieve_email(attributes, nameid, provider.config)
 
+        # Filter users based on the emails provided in the commits
+        user_emails = UserEmail.objects.filter(
```

Comment:
Change this to coerce it to a list.

So `user_emails = list(UserEmail.objects.filter(...))`

This prevents executing the query more than once.